### PR TITLE
feat: adaptive ui loading

### PR DESCRIPTION
**Pull request by BenDol**
# Adaptive UI Loading

The goal of this feature is to provide an efficient way to load the UI on the device running the client.

### How does it work?

I've introduced a new data structure for defining the users device:
```
struct Device : type->DeviceType, os->OperatingSystem
```
Using this I will dynamically load data resources from their respective directory, this is handled in two ways:

1. data resources like styles/fonts/particles can now be loaded from the `device` and/or `os` directories
```
data/
   |--styles/
      |--<base otui files>       1.   loads base styles
      |--mobile/
          |--<otui files>        2.   loads mobile styles/overrides
          |--android/
             |--<otui files>     2.1. loads android styles/overrides
          |--ios/
             |--<otui files>     2.2. loads ios styles/overrides
      |--console/
          |--<otui files>        3.   loads console styles/overrides
          |--xbox/
             |--<otui files>     3.1. loads xbox styles/overrides
          |--playstation/
             |--<otui files>     3.2. loads playstation styles/overrides
```
This will load the resource files in the given order and override
```
base -> device -> os
```
2. **loading styles via the file name rather than the data directory:** this works in a similar way only it will also try to find the defined `main widget` from which ever otui file contains it, using the same priority flow as above.
```
modules
  |--game_battle
       |--battle.otui         // loads first
       |--battle.mobile.otui  // loads second, merges and takes priority over above
       |--battle.android.otui // loads last, merges and takes priority over above
```

I have provided a simple example in this pull request. With this system we can start building different user interfaces for the different devices.

Some other notable changes in this PR:
- added widget `max-size`
- added `LogFine` for more granular logging requirements
- added `m_level` to the Logger to specify what log level you want to use output (`level >= m_level -> log(...)`)
- added `onPropertyChange` lua call to `UIWidget::setProp`

I am only really looking for feedback from maintainers.

## Checklist

  - [x] Data resource loading by device/os
  - [x] File specific style loading by device/os
  - [x] Module specific loading by device/os

### CODE REVIEW

File: data/styles/mobile/10-buttons.otui [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+  $hover !disabled:
+    image-clip: 0 23 22 23
+
+  $pressed:
+    image-clip: 0 46 22 23
+    text-offset: 1 1
+
+  $disabled:
+    color: #dfdfdf88
+    opacity: 0.8
```
Comment (dudantas) [inline data/styles/mobile/10-buttons.otui]:
```suggestion
    opacity: 0.8
    
```

File: modules/client_entergame/entergame.mobile.otui [hunk 2]
```diff
@@ -0,0 +1,3 @@
+EnterGameWindow < MainWindow
+  !text: tr('Mobile: Enter Game')
+  size: 436 498
```
Comment (dudantas) [inline modules/client_entergame/entergame.mobile.otui]:
```suggestion
  size: 436 498
  
```

File: modules/corelib/table.lua [hunk 3]
```diff
@@ -197,6 +197,13 @@ function table.collect(t, func)
     return res
 end
 
+function table.insertall(t, s)
```
Comment (dudantas) [inline modules/corelib/table.lua:200]:
```suggestion
function table.insertAll(t, s)
```

File: modules/client_styles/styles.lua [hunk 4]
```diff
@@ -1,28 +1,41 @@
...
-            g_particles.importParticle('/particles/' .. file)
+    -- try load device specific resources
+    if device then
+        local devicePath = g_platform.getDeviceShortName(device.type)
+        if devicePath ~= "" then
+            table.insertall(files, importResources(dir..'/'..devicePath, type))
+        end
+        local osPath = g_platform.getOsShortName(device.os)
+        if osPath ~= "" then
+            table.insertall(files, importResources(dir..'/'..osPath, type))
```
Comment (dudantas) [inline modules/client_styles/styles.lua:36]:
```suggestion
            table.insertAll(files, importResources(dir..'/'..devicePath, type))
        end
        local osPath = g_platform.getOsShortName(device.os)
        if osPath ~= "" then
            table.insertAll(files, importResources(dir..'/'..osPath, type))
```

File: src/framework/platform/unixplatform.cpp [hunk 5]
```diff
@@ -213,14 +224,13 @@ std::string Platform::traceback(const std::string_view where, int level, int max
...
+    std::stringstream ss;
     ss << "\nat:";
     ss << "\n\t[C++]: " << where;
     return ss.str();
-}
 #endif
+}
+
 
 #endif
```
Comment (dudantas) [inline src/framework/platform/unixplatform.cpp:243]:
```suggestion
#endif

```

File: src/framework/ui/uimanager.cpp [hunk 6]
```diff
@@ -407,26 +431,89 @@ std::string UIManager::getStyleClass(const std::string_view styleName)
...
 
-        return widget;
-    } catch (stdext::exception& e) {
+        // load device styles and widget
+        auto device = g_platform.getDevice();
+        try {
+            const auto& deviceWidgetNode = loadDeviceUI(file, device.type);
+            if (deviceWidgetNode)
+                widgetNode = deviceWidgetNode;
+        } catch (stdext::exception&) {}
```
Comment (dudantas) [inline src/framework/ui/uimanager.cpp]: Wouldn't it be interesting to send something in the exception? Maybe a log reporting the error.

**Review by dudantas — COMMENTED** [review]
The contribution is very good, congratulations, I just have a few small suggestions to make.
In some places the end line is missing, I also saw an exception that is handled but has no behaviour inside it, it would be interesting to put it in my view, as it facilitates debugging.

File: modules/corelib/table.lua (hunk 3 above)
Comment (BenDol) [reply modules/corelib/table.lua:200]: `insertAll` breaks the `table` convention of lowercase method calls. We can leave it lower case until they are all changed later to match the convention I think

File: src/framework/ui/uimanager.cpp (hunk 6 above)
Comment (BenDol) [reply src/framework/ui/uimanager.cpp]: We could do a `fine` log output here, but its expected that a lot of style loading will come back with this exception. So to me it doesn't really give us much information even for debug purposes, you're better off going into the debugger.

Comment (BenDol) [reply src/framework/ui/uimanager.cpp]: However if there is an exception we don't expect we dont want to swallow it, so I will add a fine log output. Thanks for the suggestion!

**Review by tuliomagalhaes — APPROVED** [review]
Amazing work! Thanks for doing this! 🚀

**Comment by conde2** [comment]
Amazing pull request! I would suggest editting INT_MAX to a more c++ approach using std::numeric_limits<T>::max

**Comment by beats-dh** [comment]
LGTM 🚀

**Comment by BenDol** [comment]
Added supported devices to modules, now we can load modules specific to different devices.

```
Module
  name: game_cooldown
  description: Spellcooldowns
  author: OTClient team
  website: https://github.com/edubart/otclient
  sandboxed: true
  scripts: [ cooldown ]
  devices: [ desktop, console:xbox ]
  @onLoad: init()
  @onUnload: terminate()
```
In this example this module will only load on `any desktop` and `xbox` devices, if none are provided it loads on any device as usual.

**Comment by BenDol** [comment]
@mehah This is completed now. Revisions might come later, but I think its in a good state to merge if you're happy with it.

**Comment by gaamelu** [comment]
Amazing!!! Thanks for contributing.

**Pull request merged**